Having Many Environment Variables Breaks Build: 'code too large'

### Expected Behavior

Micronaut aotshould be able to handle many (150) environment variables present on the build system when executing

```
./gradlew clean assemble
```

### Actual Behaviour

Executing `./gradlew clean assemble` fails due to a 'code to large' error

```
07:11:29.774 [main] INFO  i.m.a.s.s.YamlPropertySourceGenerator - Converting application.yml into Java based configuration
07:11:29.783 [ForkJoinPool.commonPool-worker-9] INFO  i.m.a.s.s.YamlPropertySourceGenerator - Converting application.yml into Java based configuration
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at io.micronaut.aot.cli.Main.executeInIsolatedLoader(Main.java:113)
        at io.micronaut.aot.cli.Main.run(Main.java:87)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
        at picocli.CommandLine.access$1300(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at io.micronaut.aot.cli.Main.execute(Main.java:134)
        at io.micronaut.aot.cli.Main.main(Main.java:138)
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:119)
        at java.base/java.lang.reflect.Method.invoke(Method.java:577)
        at io.micronaut.aot.cli.Main.executeInIsolatedLoader(Main.java:107)
        ... 10 more
Caused by: java.lang.RuntimeException: Compilation errors:
File EnvironmentPropertiesOptimizationLoader.java, line: 14, code too large

        at io.micronaut.aot.MicronautAotOptimizer.throwCompilationError(MicronautAotOptimizer.java:240)
        at io.micronaut.aot.MicronautAotOptimizer.compileGeneratedSources(MicronautAotOptimizer.java:140)
        at io.micronaut.aot.MicronautAotOptimizer.access$200(MicronautAotOptimizer.java:98)
        at io.micronaut.aot.MicronautAotOptimizer$Runner.execute(MicronautAotOptimizer.java:344)
        at io.micronaut.aot.MicronautAotOptimizer.execute(MicronautAotOptimizer.java:202)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        ... 12 more

> Task :prepareJitOptimizations FAILED
:prepareJitOptimizations (Thread[Execution worker Thread 12,5,main]) completed. Took 2.334 secs.

FAILURE: Build failed with an exception.
```

### Steps To Reproduce

1. Create project with the following `micronaut-cli.yml`

```yaml
applicationType: default
defaultPackage: com.example
testFramework: junit
sourceLanguage: kotlin
buildTool: gradle_kotlin
features: [annotation-api, app-name, gradle, http-client, jackson-databind, junit, kotlin, kotlin-application, kotlin-build, logback, micronaut-build, netty-server, readme, shade, yaml]
```
2. Add `io.micronaut.aot` plugin to `build.gradle.kts`

```
 aot {
        version.set("1.1.0")
        optimizeServiceLoading.set(true)
        convertYamlToJava.set(true)
        precomputeOperations.set(true)
        cacheEnvironment.set(true)
        optimizeClassLoading.set(true)
        deduceEnvironment.set(true)
    }
```

4. Create Container Image with 150 not Micronaut specific ENVs based on `gradle:jdk17`
5. Execute `./gradlew clean assemble`

I also provided a [minimal example](https://github.com/fuhrmeistery/micronaut-aot-bug/)


### Environment Information

- JDK 17
- Container as build env

### Example Application

https://github.com/fuhrmeistery/micronaut-aot-bug/

### Version

3.8.4, 3.6.4

--- a/aot-std-optimizers/src/main/java/io/micronaut/aot/std/sourcegen/EnvironmentPropertiesSourceGenerator.java
+++ b/aot-std-optimizers/src/main/java/io/micronaut/aot/std/sourcegen/EnvironmentPropertiesSourceGenerator.java
@@ -15,32 +15,43 @@
  */
 package io.micronaut.aot.std.sourcegen;
 
-import com.squareup.javapoet.ClassName;
-import com.squareup.javapoet.ParameterizedTypeName;
-import io.micronaut.aot.core.AOTContext;
-import io.micronaut.aot.core.AOTModule;
-import io.micronaut.aot.core.codegen.AbstractCodeGenerator;
-import io.micronaut.context.env.CachedEnvironment;
-import io.micronaut.core.annotation.NonNull;
-import io.micronaut.core.util.EnvironmentProperties;
-
 import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
+import javax.lang.model.element.Modifier;
+
+import io.micronaut.aot.core.AOTContext;
+import io.micronaut.aot.core.AOTModule;
+import io.micronaut.aot.core.codegen.AbstractCodeGenerator;
+import io.micronaut.context.env.CachedEnvironment;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.optim.StaticOptimizations;
+import io.micronaut.core.util.EnvironmentProperties;
+
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.ParameterizedTypeName;
+import com.squareup.javapoet.TypeSpec;
+
+import static javax.lang.model.element.Modifier.PRIVATE;
+
 /**
  * A code generator which is responsible for precomputing the Micronaut property
  * names from environment variable names at build time.
  */
 @AOTModule(
-        id = EnvironmentPropertiesSourceGenerator.ID,
-        description = EnvironmentPropertiesSourceGenerator.DESCRIPTION
+    id = EnvironmentPropertiesSourceGenerator.ID,
+    description = EnvironmentPropertiesSourceGenerator.DESCRIPTION
 )
 public class EnvironmentPropertiesSourceGenerator extends AbstractCodeGenerator {
+
     public static final String ID = "precompute.environment.properties";
     public static final String DESCRIPTION = "Precomputes Micronaut configuration property keys from the current environment variables";
+
+    private static final int MAX_METHOD_SIZE = 30_000;
 
     private final Map<String, String> env;
 
@@ -54,19 +65,62 @@ public class EnvironmentPropertiesSourceGenerator extends AbstractCodeGenerator 
 
     @Override
     public void generate(@NonNull AOTContext context) {
-        context.registerStaticOptimization("EnvironmentPropertiesOptimizationLoader", EnvironmentProperties.class, initializer -> {
-            EnvironmentProperties props = EnvironmentProperties.empty();
-            env.keySet().forEach(props::findPropertyNamesForEnvironmentVariable);
 
-            initializer.addStatement("$T env = new $T()",
-                    ParameterizedTypeName.get(ClassName.get(Map.class), ClassName.get(String.class), ParameterizedTypeName.get(List.class, String.class)),
-                    ParameterizedTypeName.get(ClassName.get(HashMap.class), ClassName.get(String.class), ParameterizedTypeName.get(List.class, String.class)));
-            for (Map.Entry<String, List<String>> entry : props.asMap().entrySet()) {
-                String values = entry.getValue().stream().map(e -> "\"" + e + "\"").collect(Collectors.joining(", "));
-                initializer.addStatement("env.put($S, $T.asList($L))", entry.getKey(), Arrays.class, values);
+        String className = "EnvironmentPropertiesOptimizationLoader";
+
+        EnvironmentProperties props = EnvironmentProperties.empty();
+        env.keySet().forEach(props::findPropertyNamesForEnvironmentVariable);
+
+        MethodSpec.Builder mainLoadMethodBuilder = MethodSpec.methodBuilder("load")
+            .addModifiers(Modifier.PUBLIC)
+            .addAnnotation(Override.class)
+            .returns(EnvironmentProperties.class)
+            .addStatement("$T env = new $T()",
+                ParameterizedTypeName.get(ClassName.get(Map.class), ClassName.get(String.class), ParameterizedTypeName.get(List.class, String.class)),
+                ParameterizedTypeName.get(ClassName.get(HashMap.class), ClassName.get(String.class), ParameterizedTypeName.get(List.class, String.class)));
+
+        TypeSpec.Builder generatedTypeBuilder = TypeSpec.classBuilder(className)
+            .addModifiers(Modifier.PUBLIC)
+            .addSuperinterface(ParameterizedTypeName.get(StaticOptimizations.Loader.class, EnvironmentProperties.class));
+
+        int fixedLineSize = 37;
+        int methodSize = 0;
+        int index = 0;
+
+        MethodSpec.Builder loadMethodBuilder = createMethodBuilder(index);
+
+        for (Map.Entry<String, List<String>> entry : props.asMap().entrySet()) {
+            String values = entry.getValue().stream().map(e -> "\"" + e + "\"").collect(Collectors.joining(", "));
+            methodSize += fixedLineSize + entry.getKey().length() + values.length() + 2;
+            if (methodSize > MAX_METHOD_SIZE) {
+                generatedTypeBuilder.addMethod(loadMethodBuilder.build());
+                mainLoadMethodBuilder.addCode("load" + index + "(env);\n");
+                methodSize = 0;
+                index++;
+                loadMethodBuilder = createMethodBuilder(index);
             }
-            initializer.addStatement("return $T.of(env)", EnvironmentProperties.class);
-        });
+            loadMethodBuilder.addStatement("env.put($S, $T.asList($L))", entry.getKey(), Arrays.class, values);
+        }
+
+        generatedTypeBuilder.addMethod(loadMethodBuilder.build());
+        mainLoadMethodBuilder.addCode("load" + index + "(env);\n")
+            .addStatement("return $T.of(env)", EnvironmentProperties.class);
+        generatedTypeBuilder.addMethod(mainLoadMethodBuilder.build());
+
+        context.registerGeneratedSourceFile(context.javaFile(generatedTypeBuilder.build()));
+        context.registerServiceImplementation(StaticOptimizations.Loader.class, className);
     }
 
+    /**
+     * Creates default inner loader method.
+     *
+     * @param index number on loader method
+     * @return method builder
+     */
+    MethodSpec.Builder createMethodBuilder(int index) {
+        return MethodSpec.methodBuilder("load" + index)
+            .addParameter(ParameterizedTypeName.get(ClassName.get(Map.class), ClassName.get(String.class), ParameterizedTypeName.get(List.class, String.class)),
+                "env")
+            .addModifiers(PRIVATE);
+    }
 }

--- a/aot-std-optimizers/src/test/groovy/io/micronaut/aot/std/sourcegen/EnvironmentPropertiesSourceGeneratorTest.groovy
+++ b/aot-std-optimizers/src/test/groovy/io/micronaut/aot/std/sourcegen/EnvironmentPropertiesSourceGeneratorTest.groovy
@@ -49,11 +49,15 @@ import java.util.List;
 import java.util.Map;
 
 public class EnvironmentPropertiesOptimizationLoader implements StaticOptimizations.Loader<EnvironmentProperties> {
+  private void load0(Map<String, List<String>> env) {
+    env.put("MICRONAUT_PORT", Arrays.asList("micronaut.port", "micronaut-port"));
+    env.put("SOME_LONG_ENV_VAR", Arrays.asList("some.long.env.var", "some.long.env-var", "some.long-env.var", "some.long-env-var", "some-long.env.var", "some-long.env-var", "some-long-env.var", "some-long-env-var"));
+  }
+
   @Override
   public EnvironmentProperties load() {
     Map<String, List<String>> env = new HashMap<String, List<String>>();
-    env.put("MICRONAUT_PORT", Arrays.asList("micronaut.port", "micronaut-port"));
-    env.put("SOME_LONG_ENV_VAR", Arrays.asList("some.long.env.var", "some.long.env-var", "some.long-env.var", "some.long-env-var", "some-long.env.var", "some-long.env-var", "some-long-env.var", "some-long-env-var"));
+    load0(env);
     return EnvironmentProperties.of(env);
   }
 }"""

